Increase maximum replicas count in cluster to 64.

amdei opened the pull request:
Current limitation of 32 replicas in cluster is too low.
This patch increases it to 99.
Why not more? Due to comment in

./src/box/vclock.h:60:   *    <server_id> is 0..VCLOCK_MAX (2 chars),

So there is only 2 chars reserved for server ID.

Testing done:

    mkdir ~/tarantool/bin
    ln /usr/bin/python ~/tarantool/bin/python
    cd ~/tarantool/test
    PATH=~/tarantool/bin:$PATH ./test-run.py

All tests passed.
Plus custom tests ensuring that master-master replication with 99 replication_source really works.
(requires a lot of memory)

rtsisyk commented:
`2 chars` is actually `2 bytes`, so 128 is more appropriate value than 99. 
Anyway, there were some good reasons to set this limit:

1. `struct vclock` has 32 bit `unsigned int map` used to speed up comparison and iteration over vclock
2. `struct vclock` is statically allocated and huge values of VCLOCK_MAX may cause stack overlow
3. Replication relay is not optimized for the large number of replicas, see #1270

Please try to deal with the first problem before increasing VCLOCK_MAX limit.

amdei commented:
@rtsisyk Thanks for fedback!
Indeed, I overlooked vlock->map member and its purpose.

Changed VCLOCK_MAX down to 64.
Primary reason is absence of safe analogue of __builtin_popcount for 128-bit integers (for vclock_size() implementation).

Looking at the code I do not understand how my test with 99 replicas works. It looks for me that it shouldn't. Will double-check.

Thus:

   1. Fixed. Does it looks any better?
   1. Size of `struct vclock` now is 544 bytes (without alignment), and there are no arrays of vclock allocations.

As I understand my changes would break compatibility with existing XLOG files.
Is it a real issue for existing users?

kostja commented:
Sorry, this has been hanging around for too long. Please join our telegram chat to figure out what's wrong with this PR. I must close it now (please open a new PR afterwards).
